File: ccms/src/svc/directory/cms_data_value_entry.c
#include "svc/directory/cms_data_value_entry.h"
#include "data/common/cms_sub_reference.h"
#include "data/choice/cms_data.h"

int cms_data_value_entry_encode_stream(per_stream_t *s, const cms_data_value_entry_t *v) {
    if (!v || !v->reference || !v->value) return CMS_ERR;
    int err;
    err = cms_sub_reference_encode_stream(s, v->reference);
    if (err) return err;
    err = cms_data_encode_stream(s, v->value);
    if (err) return err;
    return CMS_OK;
}

int cms_data_value_entry_decode_stream(per_stream_t *s, cms_data_value_entry_t *v) {
    if (!v || !v->reference || !v->value) return CMS_ERR;
    int err;
    err = cms_sub_reference_decode_stream(s, v->reference);
    if (err) return err;
    err = cms_data_decode_stream(s, v->value);
    if (err) return err;
    return CMS_OK;
}

File: jcms/src/main/java/com/ysh/jcms/svc/directory/CmsDataValueEntry.java
package com.ysh.jcms.svc.directory;

import com.ysh.jcms.core.CmsType;
import com.ysh.jcms.data.choice.CmsData;
import com.ysh.jcms.data.common.CmsSubReference;
import java.util.Arrays;
import java.util.List;

/**
 * DataValueEntry ::= SEQUENCE {
 *     reference     [0] IMPLICIT SubReference,
 *     value         [1] IMPLICIT Data
 * }  —  8.3.4
 *
 * Used by GetAllDataValues response (SEQUENCE OF DataValueEntry).
 */
public class CmsDataValueEntry extends CmsType {

    public CmsSubReference reference;
    public CmsData         value;

    public CmsDataValueEntry() {
        this.reference = new CmsSubReference();
        this.value     = new CmsData();
    }

    @Override
    public List<? extends CmsType> children() {
        return Arrays.asList(reference, value);
    }
}
